File: day07/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"slices"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

type (
	CardHand struct {
		Kind   int8
		KindJ  int8
		Score  int
		ScoreJ int
		Bid    int
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var hands []CardHand

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		a, b, ok := strings.Cut(scanner.Text(), " ")
		if !ok || len(a) != 5 {
			log.Fatalln("Invalid line")
		}
		num, err := strconv.Atoi(b)
		if err != nil {
			log.Fatalln(err)
		}
		ab := []byte(a)
		hands = append(hands, CardHand{
			Kind:   handKind(ab, false),
			KindJ:  handKind(ab, true),
			Score:  scoreHand(ab, false),
			ScoreJ: scoreHand(ab, true),
			Bid:    num,
		})
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	slices.SortFunc(hands, func(a, b CardHand) int {
		if a.Kind == b.Kind {
			return a.Score - b.Score
		}
		return int(a.Kind - b.Kind)
	})

	sum := 0
	for n, i := range hands {
		sum += (n + 1) * i.Bid
	}
	fmt.Println("Part 1:", sum)

	slices.SortFunc(hands, func(a, b CardHand) int {
		if a.KindJ == b.KindJ {
			return a.ScoreJ - b.ScoreJ
		}
		return int(a.KindJ - b.KindJ)
	})

	sum = 0
	for n, i := range hands {
		sum += (n + 1) * i.Bid
	}
	fmt.Println("Part 2:", sum)
}

func scoreCard(b byte, withJoker bool) int {
	if b >= '2' && b <= '9' {
		return int(b - '2' + 1)
	}
	switch b {
	case 'T':
		return 9
	case 'J':
		if withJoker {
			return 0
		} else {
			return 10
		}
	case 'Q':
		return 11
	case 'K':
		return 12
	case 'A':
		return 13
	}
	return 0
}

func scoreHand(b []byte, withJoker bool) int {
	sum := 0
	for _, i := range b {
		sum = sum*14 + scoreCard(i, withJoker)
	}
	return sum
}

func handKind(b []byte, withJoker bool) int8 {
	cardCountsMap := map[byte]int8{}
	var jokers int8 = 0
	for _, i := range b {
		if withJoker && i == 'J' {
			jokers++
			continue
		}
		cardCountsMap[i] = cardCountsMap[i] + 1
	}
	if len(cardCountsMap) == 0 {
		return 6
	}
	cardCounts := make([]int8, 0, len(cardCountsMap))
	for _, v := range cardCountsMap {
		cardCounts = append(cardCounts, v)
	}
	slices.Sort(cardCounts)
	slices.Reverse(cardCounts)
	cardCounts[0] += jokers
	if cardCounts[0] == 5 {
		// 5 of kind
		return 6
	}
	if cardCounts[0] == 4 {
		// 4 of kind
		return 5
	}
	if cardCounts[0] == 3 {
		if cardCounts[1] == 2 {
			// full house
			return 4
		} else {
			// 3 of kind
			return 3
		}
	}
	if cardCounts[0] == 2 {
		if cardCounts[1] == 2 {
			// two pair
			return 2
		} else {
			// one pair
			return 1
		}
	}
	return 0
}

File: day07/src/main.rs
use std::cmp::Ordering;
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

struct CardHand {
    kind: u8,
    kind_j: u8,
    score: u64,
    score_j: u64,
    bid: usize,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut hands = reader
        .lines()
        .map(|line| {
            let line = line?;
            let (a, b) = line.split_once(" ").ok_or("Invalid line")?;
            if a.len() != 5 {
                return Err("Invalid hand".into());
            }
            let bid = b.parse::<usize>()?;
            let a = a.as_bytes();
            Ok::<_, Box<dyn std::error::Error>>(CardHand {
                kind: hand_kind(a, false),
                kind_j: hand_kind(a, true),
                score: score_hand(a, false),
                score_j: score_hand(a, true),
                bid,
            })
        })
        .collect::<Result<Vec<_>, _>>()?;

    hands.sort_unstable_by(|a, b| match a.kind.cmp(&b.kind) {
        Ordering::Equal => a.score.cmp(&b.score),
        v => v,
    });
    println!(
        "Part 1: {}",
        hands
            .iter()
            .enumerate()
            .fold(0, |acc, (i, v)| acc + (i + 1) * v.bid)
    );

    hands.sort_unstable_by(|a, b| match a.kind_j.cmp(&b.kind_j) {
        Ordering::Equal => a.score_j.cmp(&b.score_j),
        v => v,
    });
    println!(
        "Part 2: {}",
        hands
            .iter()
            .enumerate()
            .fold(0, |acc, (i, v)| acc + (i + 1) * v.bid)
    );

    Ok(())
}

fn score_card(b: u8, with_joker: bool) -> u64 {
    match b {
        b'T' => 9,
        b'J' => {
            if with_joker {
                0
            } else {
                10
            }
        }
        b'Q' => 11,
        b'K' => 12,
        b'A' => 13,
        b if b >= b'2' && b <= b'9' => (b - b'2' + 1) as u64,
        _ => 0,
    }
}

fn hand_kind(b: &[u8], with_joker: bool) -> u8 {
    let (card_counts, jokers) = b
        .iter()
        .fold((HashMap::new(), 0u8), |(mut acc, jokers), &v| {
            if with_joker && v == b'J' {
                return (acc, jokers + 1);
            }
            let e = acc.entry(v).or_insert(0u8);
            *e += 1;
            (acc, jokers)
        });
    if card_counts.len() == 0 {
        return 6;
    }
    let mut card_counts = card_counts.into_iter().map(|v| v.1).collect::<Vec<_>>();
    card_counts.sort_unstable_by(|a, b| b.cmp(a));
    card_counts[0] += jokers;

    if card_counts[0] == 5 {
        // 5 of kind
        6
    } else if card_counts[0] == 4 {
        // 4 of kind
        5
    } else if card_counts[0] == 3 {
        if card_counts[1] == 2 {
            // full house
            4
        } else {
            // 3 of kind
            3
        }
    } else if card_counts[0] == 2 {
        if card_counts[1] == 2 {
            // two pair
            2
        } else {
            1
        }
    } else {
        0
    }
}

fn score_hand(b: &[u8], with_joker: bool) -> u64 {
    b.iter()
        .fold(0, |acc, &v| acc * 14 + score_card(v, with_joker))
}
